Have as_tensor always return a float64 tensor in dynamo

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138598
* #138595

As discussed with @ezyang, this set of diffs are extracting fixes to problems discovered to flipping `specialize_float=False` in https://github.com/pytorch/pytorch/pull/137782. Since these codepaths are exercised in existing tests, I'm going to bias towards shipping speed and put these up with the primary test plan as the global CI. These code paths are all tested via existing tests when `specialize_float=False` and it feels a bit wonky to add more gated tests that only test behavior when this flag is True, especially since these code paths are already covered. That being said, I'm happy to add individual tests if reviewers insist or have a different POV.  

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

bobrenjc93 committed:
Have as_tensor always return a float64 tensor in dynamo

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1150,7 +1150,9 @@ def as_tensor(self, tx):
                 tx, torch.scalar_tensor
-            ).call_function(tx, [self], {})
+            ).call_function(
+                tx, [self], {"dtype": VariableTracker.build(tx, torch.float64)}
```
You can't do it this way, right? Need to handle int and bool too

bobrenjc93 replied:
I must have misunderstood your comment here: https://github.com/pytorch/pytorch/pull/137782/files#r1809259927

I thought you meant this code path would only be exercised by symfloats so therefore the conditional was unnecessary. Now I'm curious, what did you mean by the comment? Did you mean to enumerate all the possible dtypes?

I'll push up a new commit with the conditionals added back and add a test.

ezyang replied:
In your old PR you specifically tested for float and only passed dtype arg when it was a float. My suggestion was that you should always compute the right dtype (whether it is float or not) and always pass it in, instead of ever relying on type inference

bobrenjc93 replied:
Gotcha that makes sense. I've updated the PR accordingly.

ezyang reviewed (commented):
Test too!

bobrenjc93 committed:
Update on "Have as_tensor always return a float64 tensor in dynamo"


As discussed with ezyang, this set of diffs are extracting fixes to problems discovered to flipping `specialize_float=False` in https://github.com/pytorch/pytorch/pull/137782. Since these codepaths are exercised in existing tests, I'm going to bias towards shipping speed and put these up with the primary test plan as the global CI. These code paths are all tested via existing tests when `specialize_float=False` and it feels a bit wonky to add more gated tests that only test behavior when this flag is True, especially since these code paths are already covered. That being said, I'm happy to add individual tests if reviewers insist or have a different POV.  

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Have as_tensor always return a float64 tensor in dynamo"


As discussed with ezyang, this set of diffs are extracting fixes to problems discovered to flipping `specialize_float=False` in https://github.com/pytorch/pytorch/pull/137782. Since these codepaths are exercised in existing tests, I'm going to bias towards shipping speed and put these up with the primary test plan as the global CI. These code paths are all tested via existing tests when `specialize_float=False` and it feels a bit wonky to add more gated tests that only test behavior when this flag is True, especially since these code paths are already covered. That being said, I'm happy to add individual tests if reviewers insist or have a different POV.  

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

bobrenjc93 commented on test/dynamo/test_unspec.py:
```diff
@@ -601,6 +601,20 @@ def fn(x):
+        cf = torch.compile(backend="eager", fullgraph=True)(f)
+        x = 1.1234567891234568
+        y = 1.1234567891234569
+        self.assertAlmostEqual(f(x, y), cf(x, y))
```
I was shocked to find that if I did `assertEqual` here instead of `assertAlmostEqual`, the test would erroneously pass without my fix. I still haven't found a good explanation as to why, since it seems like under the hood `assertEqual` should use `==`.

If you print out `print(f(x, y), cf(x, y))` you will see they differ.

If you assertTrue(f(x, y) == cf(x, y)) it will fail

But if you `assertEqual(f(x, y), cf(x, y))` it will pass...

Reviewers, if you happen to know and could enlighten me on what's going on, that'd be much appreciated!

ezyang replied:
assertEqual has built in rtol/atol. If you want to test bitwise equivalence use frexp

bobrenjc93 committed:
Update on "Have as_tensor always return a float64 tensor in dynamo"


As discussed with ezyang, this set of diffs are extracting fixes to problems discovered to flipping `specialize_float=False` in https://github.com/pytorch/pytorch/pull/137782. Since these codepaths are exercised in existing tests, I'm going to bias towards shipping speed and put these up with the primary test plan as the global CI. These code paths are all tested via existing tests when `specialize_float=False` and it feels a bit wonky to add more gated tests that only test behavior when this flag is True, especially since these code paths are already covered. That being said, I'm happy to add individual tests if reviewers insist or have a different POV.  

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

ezyang reviewed (approved):
sure

bobrenjc93 commented:
@pytorchbot merge -f "ci failures unrelated"
